### ISSUE TITLE
If one provides a helper function from another library, one should avoid simply copy-pasting their code

### ISSUE THREAD

**Issue by puellanivis** [pkg/sftp#188]
If one uses the following:

```
func Match(pattern, name string) (matched bool, err error) {
  return path.Match(pattern, name)
}
```

I mean… code reuse is a good thing, right? Plus, then one can’t really take responsible for bug-fixing the code.

**Comment by eikenb** [comment]
Hadn't even looked at that file. It does look like they pretty much just copied filepath/match.go into here with a couple changes. The commit when it was added (its only commit) just says "ported Glob from filepath".

For Match() they just took the filepath version and changed it to hardcode the separator as '/', which is almost exactly what path.Match() does. So it makes some sense to  swap it for a call to path.Match().

ScanChunk() behaves just like the one in filepath.Scanchunk() as well, except the hardcoded separator and missing a windows check. In this case path.Scanchunk() is not the same, missing an error case in the handling of '[' that is also present in filepath.Scanchunk().  Wonder if this might be a bug in path.Scanchunk()?

The glob code has no path/match.go equivalent. It mostly looks like the filepath code with the windows bits stripped out.

Finally there is Join() at the end which mirrors the filepath code in structure, with the separator hardcoded, and it is missing the Clean() call which was . Hardcoding the separator makes it equivalent to the one in path/path.go. It looks like they removed the Clean() call, as it has been there in the filepath version from the start. I wonder why they removed it.

Anyways, I think I'll try changing both Match() and Join() to just call their path versions and see how it does on the test suite.

**Comment by eikenb** [comment]
Hmm. Found one reason for Match() being in this file. All the returned cases of ErrBadPattern. That is a public error and it doesn't match against path.ErrBadPattern. But if I change the definition of ErrBadPattern in our match.go to "var ErrBadPattern = path.ErrBadPattern" then it works and all mentions of ErrBadPattern would be the path version.

This has 2 ramifications.
1) If anyone depended on this ErrBadPattern being different than the one from path, it would break those tests.
2) It would make the local ErrBadPattern dependent on the exported error of another package, which is usually considered to be a bad practice.

So I'm not sure about changing this or not. Do you have any thoughts?

**Comment by puellanivis** [comment]
The large number of io.EOF returns from other libraries demonstrates that sometimes, using an exported error from another library is appropriate. I would also point to http.Status___ not exactly the same thing, but related.

Sometimes, a sentinel error is appropriate…

However, looking at the tests themselves, it seems like the tests are designed with a fault, in that they are expecting a _specific_ error, rather than an nonspecific error. How much one can avoid programmers from using Sentinel Errors when not appropriate… of obvious relevance on the topic: https://dave.cheney.net/2014/12/24/inspecting-errors

I mean, saying, “it returns an error, and don’t expect any error to be anything specific unless absolutely contractually declared” works great for pedantic people like me… but new-comers are of course prone to making assumptions that are covered under community-known best-practices but-not-codified-anywhere-official…

I think “breaking” any code that assumes that sftp.ErrBadPattern is not the same as path.ErrBadPattern is permissible in the same way that the c-library broke a bunch of programs that assumed the fread/fwrite/fclose functions would just no-op when passed a NULL filepointer… if you rely upon undefined behavior, that’s setting yourself up to be bitten.

Which, is kind of where I run into the trouble with writing tests to test otherwise unspecified behavior… why should one test any particular bad pattern to return a _specific_ error, without that error being a deeply embedded part of the contract? (i.e. io.EOF is deeply embedded, path.ErrBadPattern not so much)

Looking through the golang path/filepath code… omg, they both spin off a new version of the same error… all the sentinel error details rolled into one argument right there…

**Comment by puellanivis** [comment]
P.S.: I think a benefit to “not actually the official libraries” means a situation where one can say, “ok, we’re going to make some changes that apply the 20/20-hindsight best practices into code that Go1 standard libraries maybe can’t, because some poor-practice program might have assumed something, that we can’t break now.”

**Comment by eikenb** [comment]
The fact that both path and filepath have the same error defined as separate values is the one thing that really made me pause. Thought it was a way of making sure to not have a error value be as meaningful. I just looked back through the history and there was no reason called out, they just wanted to split local OS path handling from network protocol path handling.

Given this is all network based, using path's error doesn't conflict in spirit and IMO makes more sense than having it re-applied again.

Hopefully the change won't break anyone's code, but we haven't had an official release yet (though I'm trying to work up to that) so we aren't breaking any sort of semantic versioning contract. And it should be a fairly small chance.

TLDR; I think I'll go with my original idea of replacing Match() and Join() with the versions from path.

Note that replacing Join() breaks a few tests because of the addition of the Clean() on the returned path ('./match.go' changes to 'match.go') . It shouldn't affect functionality as they are functionally equivalent when using them as the path to the file. So I think it is OK.
